sources/packages/backend/libs/services/src/notifications/notification/notification.service.ts

```diff
@@ -70,57 +84,151 @@ export class NotificationService extends RecordDataModelService<Notification> {
-   * to execute the command as part of an existing transaction. If not provided
-   * the local repository will be used instead.
-   * @returns GC Notify API call response.
+   * Process all the unsent notifications with a polling limit.
+   * Processing continues recursively until all the records are processed.
+   * Call GCNotify to send email notification.
+   * @param pollingRecordsLimit Maximum number of notifications to be processed in one chunk of
+   * processing.
+   * @returns processing summary.
    */
-  async sendEmailNotification(
-    notificationId: number,
-    entityManager?: EntityManager,
-  ): Promise<GCNotifyResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    const notification = await notificationRepo.findOne({
+  async processUnsentNotifications(
+    pollingRecordsLimit: number,
+  ): Promise<NotificationProcessingSummary> {
+    return await this.processUnsentNotificationsRecursive(pollingRecordsLimit);
+  }
+
+  /**
+   * Call GC Notify to process the given notification.
+   * @param notification notification to be processed.
+   * @returns status of notification processing.
+   */
+  private async sendEmailNotification(
+    notification: Notification,
+  ): Promise<boolean> {
+    // Call GC Notify send email method.
+    try {
+      await this.gcNotifyService.sendEmailNotification(
+        notification.messagePayload,
+      );
+      await this.updateNotification(notification.id);
+      return true;
+    } catch (error: unknown) {
+      this.logger.error(`Error while processing notification: ${error}`);
+      if (
+        error instanceof CustomNamedError &&
+        error.name === GC_NOTIFY_PERMANENT_FAILURE_ERROR
+      ) {
+        const processNotificationError =
+          error.objectInfo as GCNotifyErrorResponse;
+        await this.updateNotification(
+          notification.id,
+          processNotificationError.errors,
+        );
+        return false;
+      }
+      throw error;
+    }
+  }
+
+  /**
+   * Process all the unsent notifications with a polling limit recursively.
+   * Processing continues recursively until all the records are processed.
+   * Call GCNotify to send email notification.
+   * @param pollingLimit
+   * @param notificationsProcessed optional param used to
+   * increment count on recursion.
+   * @param notificationsSuccessfullyProcessed optional param used to
+   * increment count on recursion.
+   * @returns processing summary.
+   */
+  private async processUnsentNotificationsRecursive(
+    pollingRecordsLimit: number,
+    notificationsProcessed = 0,
+    notificationsSuccessfullyProcessed = 0,
+  ): Promise<NotificationProcessingSummary> {
+    const notificationsToProcess = await this.repo.find({
       select: {
         id: true,
         messagePayload: true,
       },
       where: {
-        id: notificationId,
+        dateSent: IsNull(),
```

Comment:
Can we create a tech debt t have an index added to the table that includes "dateSent/createdAt"?